livingdoc-engine/src/main/kotlin/org/livingdoc/engine/execution/examples/scenarios/ScenarioExecution.kt

```diff
@@ -0,0 +1,153 @@
+import org.livingdoc.engine.execution.examples.scenarios.model.StepResult
+import org.livingdoc.engine.fixtures.FixtureMethodInvoker
+import org.livingdoc.engine.fixtures.FixtureMethodInvoker.FixtureMethodInvocationException
+import java.lang.reflect.Parameter
+
+internal class ScenarioExecution(
+        private val fixtureClass: Class<*>,
+        scenario: Scenario,
+        document: Any?
+) {
+
+    private val fixtureModel = ScenarioFixtureModel(fixtureClass)
+
+    private val scenario = ScenarioResult.from(scenario)
+
+    private val methodInvoker = FixtureMethodInvoker(document)
+
+    /**
+     * Executes the configured [Scenario].
+     *
+     * Does not throw any kind of exception. Exceptional state of the execution is packaged inside the [ScenarioResult] in
+     * the form of different result objects.
+     */
+    fun execute(): ScenarioResult {
+        try {
+            assertFixtureIsDefinedCorrectly()
+            executeScenario()
+            markScenarioAsSuccessfullyExecuted()
+        } catch (e: Throwable) {
+            markScenarioAsExecutedWithException(e)
+        }
+        setSkippedStatusForAllUnknownResults()
+        return scenario
+    }
+
+    private fun assertFixtureIsDefinedCorrectly() {
+        val errors = ScenarioFixtureChecker.check(fixtureModel)
+        if (errors.isNotEmpty()) {
+            throw MalformedScenarioFixtureException(fixtureClass, errors)
+        }
+    }
+
+    private fun executeScenario() {
+        val fixture = createFixtureInstance()
+        executeWithBeforeAndAfter(
+                before = { invokeBeforeMethods(fixture) },
+                body = { executeSteps(fixture) },
+                after = { invokeAfterMethods(fixture) }
+        )
+    }
+
+    private fun executeSteps(fixture: Any) {
+        var previousResult: Result = Result.Executed
+        for (step in scenario.steps) {
+            if (previousResult == Result.Executed) {
+                executeStep(fixture, step)
+                previousResult = step.result
+            } else {
+                step.result = Result.Skipped
+            }
+        }
+    }
+
+    private fun executeStep(fixture: Any, step: StepResult) {
+        val result = fixtureModel.getMatchingStepTemplate(step.value)
+        val method = fixtureModel.getStepMethod(result.template)
+        val parameterList = method.parameters
+                .map {
+                    result.variables.getOrElse(
+                            getParameterName(it),
+                            { error("Missing parameter value: ${getParameterName(it)}") })
+                }
+                .toTypedArray()
+        step.result = invokeExpectingException {
+            methodInvoker.invoke(method, fixture, parameterList)
+        }
+    }
+
+    private fun getParameterName(parameter: Parameter): String {
+        return parameter.getAnnotationsByType(Binding::class.java).firstOrNull()?.value
```

Comment:
Shouldn't there be some code checking whether the parameter actually has a name? in case the `-parameter` compile flag is not applied, there will be an exception stating that no matching parameter was found, but currently the developer is not told to use either the `@Binding` annotation or compile his code with the `-parameter` flag.

As far as I know the `Parameter` reflection class in Java 8 offers a method for determining if an actual name is present.

Reply:
maybe even do this part in the `ScenarioFixtureChecker`

Reply:
I've extracted one more thing to do into #87